rosbag2_py/test/test_transport.py

```diff
@@ -40,3 +47,49 @@ def test_options_qos_conversion():
     record_options = rosbag2_py.RecordOptions()
     record_options.topic_qos_profile_overrides = simple_overrides
     assert record_options.topic_qos_profile_overrides == simple_overrides
+
+
+def test_record_cancel(tmp_path):
+    bag_path = str(tmp_path / 'test_record_cancel')
+    storage_options, converter_options = get_rosbag_options(bag_path)
+
+    recorder = rosbag2_py.Recorder()
+
+    record_options = rosbag2_py.RecordOptions()
+    record_options.all = True
+    record_options.is_discovery_disabled = False
+    record_options.rmw_serialization_format = ""
+    record_options.topic_polling_interval = datetime.timedelta(milliseconds=100)
+
+    rclpy.init()
+    record_thread = threading.Thread(
+        target=recorder.record,
+        args=(storage_options, record_options),
+        daemon=True)
+    record_thread.start()
+
+    node = rclpy.create_node('test_record_cancel')
+    executor = rclpy.executors.SingleThreadedExecutor()
+    executor.add_node(node)
+    pub = node.create_publisher(String, 'chatter', 10)
+
+    i = 0
+    msg = String()
+
+    while rclpy.ok() and i < 10:
+        msg.data = 'Hello World: {0}'.format(i)
+        i += 1
+        pub.publish(msg)
+
+    recorder.cancel()
+
+    time_counter = 0
```

Comment:
This is a little bit of an odd way to do timeout waiting - generally you want to add a timeout in terms of time, rather than a "number of times to sleep". In `rosbag2_test_common` we have a utility function `spin_and_wait_for` - https://github.com/ros2/rosbag2/blob/master/rosbag2_test_common/include/rosbag2_test_common/wait_for.hpp#L25

I think it would make sense to give yourself a utility function in `common.py` called simply `wait_for`, that could look like this:

```
def wait_for(condition, timeout: rclpy.Duration):
  clock = rclpy.Clock(RCL_STEADY_TIME)
  start = clock.now()
  while !condition():
    if clock.now() - start > timeout:
      return false
    # could make this configurable as well if you like - it's just a rate throttle so this loop doesn't eat full CPU
    time.sleep(0.1)
  return true
```

Then you could write the test as nice and simple, like:

```
metadata_path = bag_path / 'metadata.yaml'
db3_path = bag_path /  'test_record_cancel_0.db3'
assert wait_for(lambda: metadata_path.is_file() and db3_path.is_file(), timeout=rclpy.Duration(seconds=3)) 
```

That keeps the implementation details out and makes the test very self-documenting